platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/serialize/ClassLoading.kt

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.cc.impl.serialize
+
+
+private
+val defaultClassLoader = DefaultClassEncoder::class.java.classLoader
+
+
+internal
+fun classForName(name: String, classLoader: ClassLoader?): Class<*> =
```

Comment:
🤔 I'm starting to find our handling of `null` classloader quite confusing. `null` is a bootstrap classloader, but we're using it as a placeholder for the Gradle runtime classloader. I'd prefer us spelling out runtime classloader explicitly:
* Change `decodeClassLoader` to return non-`null` runtime classloader when loading an unscoped classloader.
* Forbid this function to accept a `null` classloader and get rid of substitution there.
* Introduce a `hasClassInRuntimeClassloader(name:String): Boolean` with the implementation like
```kotlin
try {
  Class.forName(name, defaultClassLoader)
  return true
} catch (e: ClassNotFoundException) {
  return false
}
```
and use it to implement `ensureClassCanBeFoundInDefaultClassLoader`.

Reply:
I moved in that direction but there's still the `Proxy` case to consider so I'm leaving the `decodeClassLoader` change to a future PR.